golang/cosmos/x/vtransfer/types/baseaddr.go

```diff
@@ -0,0 +1,153 @@
+package types
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+
+	transfertypes "github.com/cosmos/ibc-go/v6/modules/apps/transfer/types"
+	clienttypes "github.com/cosmos/ibc-go/v6/modules/core/02-client/types"
+	channeltypes "github.com/cosmos/ibc-go/v6/modules/core/04-channel/types"
+	ibcexported "github.com/cosmos/ibc-go/v6/modules/core/exported"
+)
+
+type AddressRole string
+
+const (
+	RoleSender   AddressRole = "Sender"
+	RoleReceiver AddressRole = "Receiver"
+)
+
+func trimSlashPrefix(s string) string {
+	return strings.TrimPrefix(s, "/")
+}
+
+// ExtractBaseAddress extracts the base address from a parameterized address.
+// It removes all subpath and query components from addr.
+func ExtractBaseAddress(addr string) (string, error) {
+	parsed, err := url.Parse(addr)
+	if err != nil {
+		return "", err
+	}
+
+	// Specify the fields and values we expect.  Unspecified fields will only
+	// match if they are zero values in order to be robust against extensions to
+	// the url.URL struct.
+	//
+	// Remove leading slashes from the path fields so that only parsed relative
+	// paths match the expected test.
+	expected := url.URL{
+		Path:        trimSlashPrefix(parsed.Path),
+		RawPath:     trimSlashPrefix(parsed.RawPath),
+		RawQuery:    parsed.RawQuery,
+		Fragment:    parsed.Fragment,
+		RawFragment: parsed.RawFragment,
+
+		// Skip over parsing control flags.
+		ForceQuery: parsed.ForceQuery,
+		OmitHost:   parsed.OmitHost,
+	}
+
+	if *parsed != expected {
+		return "", fmt.Errorf("address must be relative path with optional query and fragment, got %s", addr)
+	}
+
+	baseAddr, _, _ := strings.Cut(expected.Path, "/")
+	if baseAddr == "" {
+		return "", fmt.Errorf("base address cannot be empty")
+	}
+
+	return baseAddr, nil
+}
+
+// ExtractBaseTransferData returns the base address from the transferData.Sender
+// (if RoleSender) or transferData.Receiver (if RoleReceiver). Errors in
+// determining the base address are ignored... we then assume the base address
+// is exactly the original address.  If newTransferData is not nil, it will be
+// populated with a new FungibleTokenPacketData consisting of the role replaced
+// with its base address.
+func ExtractBaseTransferData(transferData transfertypes.FungibleTokenPacketData, role AddressRole, newTransferData *transfertypes.FungibleTokenPacketData) (string, error) {
```

Comment:
Exporting this function fosters unnecessary confusion w.r.t. `ExtractBaseTransferPacket` IMO.
```suggestion
// extractBaseTransferData returns the base address from the transferData.Sender
// (if RoleSender) or transferData.Receiver (if RoleReceiver). Errors in
// determining the base address are ignored... we then assume the base address
// is exactly the original address.  If newTransferData is not nil, it will be
// populated with a new FungibleTokenPacketData consisting of the role replaced
// with its base address.
func extractBaseTransferData(transferData transfertypes.FungibleTokenPacketData, role AddressRole, newTransferData *transfertypes.FungibleTokenPacketData) (string, error) {
```